lib/BoundaryValueDiffEqCore/src/utils.jl

```diff
@@ -597,3 +614,111 @@ end
 function __split_kwargs(; abstol, adaptive, controller, kwargs...)
     return ((abstol, adaptive, controller), (; abstol, adaptive, kwargs...))
 end
+
+@inline __concrete_kwargs(nlsolve, ::Nothing, nlsolve_kwargs, optimize_kwargs) = (;
+    nlsolve_kwargs...)
+@inline __concrete_kwargs(::Nothing, optimize, nlsolve_kwargs, optimize_kwargs) = (;) # Doesn't support for now
+@inline __concrete_kwargs(::Nothing, ::Nothing, nlsolve_kwargs, optimize_kwargs) = (;
+    nlsolve_kwargs...)
+
+## Optimization solver related utils ##
+
+@inline __default_cost(::Nothing) = (x, p) -> 0.0
+@inline __default_cost(f) = f
+@inline __default_cost(fun::BVPFunction) = __default_cost(fun.cost)
+
+@inline function __extract_lcons_ucons(prob::AbstractBVProblem, ::Type{T}, M, N) where {T}
+    lcons = if isnothing(prob.lcons)
+        zeros(T, N*M)
+    else
+        lcons_length = length(prob.lcons)
+        vcat(prob.lcons, zeros(T, N*M - lcons_length))
+    end
+    ucons = if isnothing(prob.ucons)
+        zeros(T, N*M)
```

Comment:
Using zero is likely not good? That will mess with the stability of the solver quite a bit. If inequality constraints are given then the lcons / ucons should be required from the user.